feat(ondemand): Add ondemand checks to the dashboard widget validation

wmak opened the pull request:
- This adds a cardinality check to when we verify dashboard widgets and a spec limit check. Returning either result as warnings on the response so the frontend can use these in displaying a toggle where ondemand can be disabled via `onDemandExtractionDisabled`

wmak committed:
feat(ondemand): Add cardinality limit check

- wip: adds a check for all the field cardinality

wmak committed:
wip

wmak committed:
hmm should be good enough

wmak committed:
??? not sure why this reverted

wmak committed:
col not fields

wmak committed:
ref: Get sepc-limit working

wmak committed:
remove onDemandExtractionEnabled flag

wmak committed:
Merge branch 'master' into wmak/feat/add-cardinality-limit-check

wmak committed:
Revert "remove onDemandExtractionEnabled flag"

This reverts commit 9bd33e25a0cee361303e768655d1bb295c466661.

wmak committed:
fix tests

wmak committed:
ref: change flag to `disabled` so we check ondemand by default

wmak committed:
use constants for the warnings

wmak committed:
s/fields/columns

k-fish commented on src/sentry/api/serializers/rest_framework/dashboard.py:
```diff
@@ -305,6 +318,40 @@ def validate(self, data):
                 else:
                     query_errors.append({})
+
+                if not query.get("on_demand_extraction_disabled", False):
```
Let's add a flag check w/ `organizations:on-demand-metrics-extraction-widgets` in addition to the passed `disabled` flag here so we're not making these for orgs that aren't extracting. If it's not set, all the specs will be wrong anyway because of https://github.com/getsentry/sentry/blob/7f3992553de326213c080a68f22033940a193ab7/src/sentry/relay/config/metric_extraction.py#L197

k-fish commented on src/sentry/api/serializers/rest_framework/dashboard.py:
```diff
@@ -305,6 +318,40 @@ def validate(self, data):
+                    # Get widget specs if we haven't yet
+                    if current_widget_specs is None:
+                        current_widget_specs = get_current_widget_specs(organization)
+                    widget_specs = _get_widget_on_demand_specs(widget_query, organization)
+                    if len(widget_specs) == 0:
+                        # In the case the query errors, we should still disbale
+                        self.query_warnings["queries"].append(
+                            OnDemandExtractionState.DISABLED_QUERY_ERROR
+                        )
```
OnDemandExtractionState.DISABLED_QUERY_ERROR -> `DISABLED_NOT_APPLICABLE` since it's an expected case that we have **0** specs; MEP data covers some queries already in which case a spec isn't created.

It's possible we don't need this either, since we should be checking again during the save whether it has specs, and if it doesn't, saving as `NOT_APPLICABLE` then too.

k-fish commented on src/sentry/models/dashboard_widget.py:
```diff
@@ -144,6 +144,8 @@ class OnDemandExtractionState(models.TextChoices):
         DISABLED_HIGH_CARDINALITY = "disabled:high-cardinality", gettext_lazy(
             "disabled:high-cardinality"
         )
+        """ This widget has a query which errors so we shouldn't extract anything """
+        DISABLED_QUERY_ERROR = "disabled:query-error", gettext_lazy("disabled:query-error")
```
I don't think we need this

k-fish commented on src/sentry/options/defaults.py:
```diff
@@ -1754,6 +1754,11 @@
 )
+register(
+    "on_demand.max_widget_cardinality.on_query_count",
+    default=100,
```
```suggestion
    default=50,
```
Let's bump this down to `50` actually, I'd prefer we start low and bump it up later if needed. Widgets will fall back to indexed data anyway.

k-fish commented on src/sentry/tasks/on_demand_metrics.py:
```diff
@@ -330,44 +335,70 @@ def _get_widget_query_low_cardinality(
     if not query_columns:
         return None
 
+    # We cache each key individually to query less
```
May want to add a rollout/killswitch here, just in case.

wmak replied:
I'll add the killswitch here: https://github.com/getsentry/sentry/pull/65051

k-fish commented on src/sentry/tasks/on_demand_metrics.py:
```diff
@@ -39,6 +39,7 @@
 
 # TTL for cardinality check
 _WIDGET_QUERY_CARDINALITY_TTL = 3600 * 24 * 7  # Cardinality outcome is valid for 7 days.
+_COLUMN_CARDINALITY_TTL = 3600 * 24  # Cardinality outcome is valid for 1 days.
```
Maybe we should make this explicitly the "save/edit" widget TTL and have a different TTL for the task? 24 hours is a long time for fields that can change when they are just introduced and someone is adding a widget.

wmak replied:
from slack; Going to change the column TTL to 1hr, and the query period for widget saves from 30m to 1hr as well

k-fish commented on tests/sentry/api/endpoints/test_organization_dashboard_widget_details.py:
```diff
@@ -665,3 +679,215 @@ def test_accepts_environment_for_filters_that_require_single_env(self):
+        assert response.status_code == 200, response.data
+        # There's no data, so `sometag` should be low cardinality
+        assert len(response.data) == 0
+        # We cache so we shouldn't call query cardinality again
+        with mock.patch("sentry.tasks.on_demand_metrics._query_cardinality") as mock_query:
+            self.client.post(f"{self.url()}?environment=mock_env", data)
+            assert len(mock_query.mock_calls) == 0
```
Do these tests need `ONDEMAND_FEATURES`? no features -> no specs -> no queries?

k-fish commented on src/sentry/tasks/on_demand_metrics.py:
```diff
@@ -330,44 +335,70 @@ def _get_widget_query_low_cardinality(
+    if len(cardinality_map) == len(query_columns):
+        return cardinality_map
+
+    query_columns = [col for col in query_columns if col not in cardinality_map]
```
This isn't right -> query_columns if col will never be in the cache key list.

wmak committed:
ref: address PR comments

wmak committed:
Merge branch 'master' into wmak/feat/add-cardinality-limit-check

wmak committed:
ref: Change caching strategy

- Cache each column individually, but in different key spaces depending
  on whether we're running the check in the task or in the pre-save
  checks

k-fish committed:
Update src/sentry/api/serializers/rest_framework/dashboard.py

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

k-fish commented on src/sentry/api/serializers/rest_framework/dashboard.py:
```diff
@@ -305,6 +326,37 @@ def validate(self, data):
+                        current_widget_specs = get_current_widget_specs(organization)
+                    widget_specs = _get_widget_on_demand_specs(widget_query, organization)
+                    if len(widget_specs) == 0:
+                        # In the case the query errors, we should still disbale
```
```suggestion
                        # Disabled since there are no applicable widgets
```

k-fish committed:
Update src/sentry/api/serializers/rest_framework/dashboard.py
